src/mod_carboncopy.erl

```diff
@@ -123,69 +126,102 @@ user_receive_packet(Acc, JID, _From, To, Packet) ->
+        <<"normal">> -> contains_body(Packet) orelse
+                        contains_receipts(Packet) orelse
+                        contains_csn(Packet);
+        <<"chat">> -> true;
+        _ -> false
     end.
 
--spec is_received(_) -> classification().
+-spec is_valid_muc(exml:element(), jid:jid(), direction()) -> boolean().
+is_valid_muc(_, _, sent) ->
+    false;
+is_valid_muc(Packet, To, _) ->
+    is_mediated_invitation(Packet) orelse
+    is_direct_muc_invitation(Packet) orelse
+    is_received_private_muc(Packet, To).
+
+-spec is_mediated_invitation(exml:element()) -> boolean().
+is_mediated_invitation(Packet) ->
+    undefined =/= exml_query:path(Packet,
+                                  [{element_with_ns, <<"x">>, ?NS_MUC_USER},
+                                   {element, <<"invite">>},
+                                   {attr, <<"from">>}]).
+
+-spec is_direct_muc_invitation(exml:element()) -> boolean().
+is_direct_muc_invitation(Packet) ->
+    undefined =/= exml_query:subelement_with_name_and_ns(Packet, <<"x">>, ?NS_CONFERENCE).
+
+-spec is_received_private_muc(exml:element(), jid:jid()) -> boolean().
+is_received_private_muc(_, #jid{lresource = <<>>}) ->
+    false;
+is_received_private_muc(Packet, _) ->
+    undefined =/= exml_query:subelement_with_name_and_ns(Packet, <<"x">>, ?NS_MUC_USER).
+
+-spec is_carbon_private(exml:element()) -> boolean().
+is_carbon_private(Packet) ->
+    undefined =/= exml_query:subelement_with_name_and_ns(Packet, <<"private">>, ?NS_CC_2).
+
+-spec has_nocopy_hint(exml:element()) -> boolean().
+has_nocopy_hint(Packet) ->
+    undefined =/= exml_query:subelement_with_name_and_ns(Packet, <<"no-copy">>, ?NS_HINTS).
+
+-spec contains_body(exml:element()) -> boolean().
+contains_body(Packet) ->
+    undefined =/= exml_query:subelement(Packet, <<"body">>).
+
+-spec contains_receipts(exml:element()) -> boolean().
+contains_receipts(Packet) ->
+    undefined =/= exml_query:subelement_with_name_and_ns(Packet, <<"received">>, ?NS_RECEIPTS).
+
+-spec contains_csn(exml:element()) -> boolean().
+contains_csn(Packet) ->
+    undefined =/= exml_query:subelement_with_ns(Packet, ?NS_CHATSTATES).
+
+-spec is_received(exml:element()) -> boolean().
 is_received(Packet) ->
-    case xml:get_subtag(Packet, <<"received">>) of
-        false -> is_sent(Packet);
-        _ -> ignore
-    end.
+    undefined =/= exml_query:subelement_with_name_and_ns(Packet, <<"received">>, ?NS_CC_2).
 
--spec is_sent(_) -> classification().
+-spec is_sent(exml:element()) -> boolean().
 is_sent(Packet) ->
-    case xml:get_subtag(Packet, <<"sent">>) of
-        false -> forward;
-        SubTag -> is_forwarded(SubTag)
-    end.
+    undefined =/= exml_query:subelement_with_name_and_ns(Packet, <<"sent">>, ?NS_CC_2).
 
--spec is_forwarded(_) -> classification().
-is_forwarded(SubTag) ->
-    case xml:get_subtag(SubTag, <<"forwarded">>) of
-        false -> forward;
-        _ -> ignore
-    end.
```

Comment:
Why is this one removed? Would it be possible to list the changed rules in a commit message, PR description or somewhere else? The refactoring makes it a bit hard to figure out which changes are intentional. According to the XEP:

> The sending server MUST NOT deliver forwarded <message/>s to the other Carbons-enabled resources of the sender.
The receiving server MUST NOT deliver forwarded <message/>s to the other Carbons-enabled resource of the recipient

Reply:
Also, as far as I understand the xep, the only way the xep proposes to have incomplete conversations across carbons-enabled devices y by the `<private/>` tag 🤔

Reply:
Ok, I trust you on this one. There are some unclear things in the XEP but in general I agree that the rules you chose look reasonable.